File: src/UnwindMC/Analysis/Analyzer.cs
using NDis86;
using NLog;
using System.Collections.Generic;
using System.Linq;
using System;
using UnwindMC.Collections;
using UnwindMC.Util;

namespace UnwindMC.Analysis
{
    public class Analyzer
    {
        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();

        private readonly InstructionGraph _graph;
        private readonly IImportResolver _importResolver;
        private readonly SortedDictionary<ulong, Function> _functions = new SortedDictionary<ulong, Function>();
        private readonly SortedDictionary<ulong, JumpTable> _jumpTables = new SortedDictionary<ulong, JumpTable>(); 

        public Analyzer(ArraySegment<byte> textBytes, ulong pc, IImportResolver importResolver)
        {
            _graph = new InstructionGraph(textBytes, pc);
            _importResolver = importResolver;
        }

        public InstructionGraph Graph => _graph;
        public IDictionary<ulong, Function> Functions => _functions;

        public void Analyze()
        {
            AddExplicitCalls();
            ResolveFunctionBounds();
            ResolveExternalFunctionCalls();
        }

        public void AddFunction(ulong address)
        {
            _functions.Add(address, new Function(address));
        }

        private void AddExplicitCalls()
        {
            Logger.Info("Adding explicit calls");
            foreach (var instr in _graph.Instructions)
            {
                if (instr.Code == MnemonicCode.Icall && instr.Operands[0].Type == OperandType.ImmediateBranch)
                {
                    var targetAddress = instr.GetTargetAddress();
                    _graph.AddLink(instr.Offset, targetAddress, InstructionGraph.LinkType.Call);
                    if (!_functions.ContainsKey(targetAddress))
                    {
                        _functions.Add(targetAddress, new Function(targetAddress));
                    }
                }
            }
            Logger.Info("Done");
        }

        private void ResolveFunctionBounds()
        {
            Logger.Info("Resolving function bounds");
            int index = 0;
            var keys = _functions.Keys;
            foreach (var key in keys)
            {
                var function = _functions[key];
                Logger.Debug("[{0}/{1}] sub_{2:x8}...", ++index, _functions.Count, function.Address);
                if (!_graph.InBounds(function.Address))
                {
                    Logger.Warn("The specified address is outside of code segment");
                    function.Status = FunctionStatus.BoundsNotResolvedInvalidAddress;
                    continue;
                }
                if (!_graph.Contains(function.Address))
                {
                    Logger.Debug("The specified address is not a valid start of instruction, re-disassembling");
                    _graph.Redisassemble(function.Address);
                }
                var visitedAllLinks = _graph
                    .WithEdgeFilter(e => (e.Type & InstructionGraph.LinkType.Next | InstructionGraph.LinkType.Branch | InstructionGraph.LinkType.SwitchCaseJump) != 0)
                    .DFS(function.Address, (instr, link) =>
                {
                    _graph.GetExtraData(instr.Offset).FunctionAddress = function.Address;
                    if (instr.Code == MnemonicCode.Iret)
                    {
                        return false;
                    }
                    AddNextLinks(instr);
                    AddExplicitBranches(instr);
                    AddSwitchCases(instr);
                    return true;
                });
                function.Status = visitedAllLinks
                    ? FunctionStatus.BoundsResolved
                    : FunctionStatus.BoundsNotResolvedIncompleteGraph;
            }
            Logger.Info("Done");
        }

        private void ResolveExternalFunctionCalls()
        {
            Logger.Info("Resolving external calls");
            foreach (var instr in _graph.Instructions)
            {
                var import = TryGetImportAddress(instr, 1)
                    .OrElse(() => TryGetImportAddress(instr, 0))
                    .Map(_importResolver.GetImportName);
                if (import.TryGet(out var name))
                {
                    _graph.GetExtraData(instr.Offset).ImportName = name;
                }
            }
            Logger.Info("Done");
        }

        private Option<ulong> TryGetImportAddress(Instruction instr, int index)
        {
            if (instr.Operands.Count > index && instr.Operands[index].Type == OperandType.Memory &&
                instr.Operands[index].Base == OperandType.None && instr.Operands[index].Index == OperandType.None &&
                _importResolver.IsImportAddress(instr.Operands[index].LValue.udword))
            {
                return Option.Some(instr.Operands[index].LValue.uqword);
            }
            return Option<ulong>.None;
        }

        private void AddNextLinks(Instruction instr)
        {
            switch (instr.Code)
            {
                case MnemonicCode.Iret:
                case MnemonicCode.Ijmp:
                case MnemonicCode.Iint3:
                    break;
                default:
                    var next = _graph.GetNext(instr.Offset);
                    if (next != _graph.FirstAddressAfterCode)
                    {
                        _graph.AddLink(instr.Offset, next, InstructionGraph.LinkType.Next);
                    }
                    break;
            }
        }

        private void AddExplicitBranches(Instruction instr)
        {
            switch (instr.Code)
            {
                case MnemonicCode.Ija:
                case MnemonicCode.Ijae:
                case MnemonicCode.Ijb:
                case MnemonicCode.Ijbe:
                case MnemonicCode.Ijcxz:
                case MnemonicCode.Ijecxz:
                case MnemonicCode.Ijg:
                case MnemonicCode.Ijge:
                case MnemonicCode.Ijl:
                case MnemonicCode.Ijle:
                case MnemonicCode.Ijmp:
                case MnemonicCode.Ijno:
                case MnemonicCode.Ijnp:
                case MnemonicCode.Ijns:
                case MnemonicCode.Ijnz:
                case MnemonicCode.Ijo:
                case MnemonicCode.Ijp:
                case MnemonicCode.Ijrcxz:
                case MnemonicCode.Ijs:
                case MnemonicCode.Ijz:
                    if (instr.Operands[0].Type == OperandType.ImmediateBranch)
                    {
                        _graph.AddLink(instr.Offset, instr.GetTargetAddress(), InstructionGraph.LinkType.Branch);
                    }
                    break;
            }
        }

        private void AddSwitchCases(Instruction instr)
        {
            if (instr.Code != MnemonicCode.Ijmp || instr.Operands[0].Type != OperandType.Memory)
            {
                return;
            }
            var address = instr.Operands[0].LValue.udword;
            if (!_jumpTables.TryGetValue(address, out var table))
            {
                table = new JumpTable(instr.Offset, address);
                ResolveJumpTable(table);
                _jumpTables.Add(address, table);
            }
            for (int i = table.FirstIndex; i < table.Count; i++)
            {
                _graph.AddLink(table.Reference, _graph.ReadUInt32(table.Address + (uint)(i * 4)), InstructionGraph.LinkType.SwitchCaseJump);
            }
        }

        private void ResolveJumpTable(JumpTable table)
        {
            if (table.Address >= _graph.FirstAddressAfterCode)
            {
                Logger.Warn("Jump table is not in the code segment");
                return;
            }

            Logger.Info("Resolving jump table tbl_{0:x8}", table.Address);
            OperandType idx = OperandType.None;
            OperandType lowByteIdx = OperandType.None;
            ulong indirectAddress = 0;
            var casesCountOption = Option<int>.None;
            _graph
                .WithEdgeFilter(e => (e.Type & InstructionGraph.LinkType.Next | InstructionGraph.LinkType.Branch) != 0)
                .ReverseEdges()
                .DFS(table.Reference, (instr, link) =>
            {
                // find out the jump index register
                if (idx == OperandType.None)
                {
                    idx = instr.Operands[0].Index;
                    lowByteIdx = RegisterHelper.GetLowByteRegisterFromDword(idx);
                    return true;
                }
                // update the main jump register if the jump is indirect
                if (indirectAddress == 0 && instr.Code == MnemonicCode.Imov && instr.Operands[0].Base == lowByteIdx)
                {
                    idx = instr.Operands[1].Base;
                    indirectAddress = instr.Operands[1].LValue.udword;
                    return true;
                }
                // search for the jump to the default case
                if (!(instr.Code == MnemonicCode.Ija && instr.Operands[0].Type == OperandType.ImmediateBranch))
                {
                    return true;
                }
                // search for cases count, can find it from code like cmp ecx, 0xb
                // the jump register is irrelevant since it must be the closest one to ja
                var tracker = new AssignmentTracker(_graph);
                var value = tracker.Find(instr.Offset, idx, (i, _) =>
                    i.Code == MnemonicCode.Icmp && i.Operands[0].Type == OperandType.Register);
                casesCountOption = value.Map(v => v.ubyte + 1);
                return false;
            });

            if (!casesCountOption.TryGet(out var casesCount))
            {
                Logger.Info("Done");
                return;
            }

            int jumpsCount;
            if (indirectAddress == 0)
            {
                jumpsCount = casesCount;
                casesCount = 0;
            }
            else
            {
                jumpsCount = _graph.GetBytes(indirectAddress, casesCount).Max() + 1;
            }

            uint offset = 0;
            for (int i = 0; i < jumpsCount; i++)
            {
                if (!_graph.AddJumpTableEntry(table.Address + offset))
                {
                    table.FirstIndex++;
                }
                _graph.AddJumpTableEntry(table.Address + offset);
                offset += 4;
            }
            table.Count = jumpsCount;
            while (casesCount >= 4)
            {
                _graph.AddJumpTableIndirectEntries(table.Address + offset, 4);
                casesCount -= 4;
                offset += 4;
            }
            if (casesCount > 0)
            {
                _graph.AddJumpTableIndirectEntries(table.Address + offset, casesCount);
                offset += (uint)casesCount;
            }
            _graph.Redisassemble(table.Address + offset);
            Logger.Info("Done");
        }
    }
}

File: src/UnwindMC/Analysis/Analyzer.fs
module rec Analyzer

open System
open System.Collections.Generic
open System.Linq
open NDis86
open NLog
open Graphs
open InstructionExtensions

type FunctionStatus =
    | Created
    | BoundsResolved
    | BoundsNotResolvedInvalidAddress
    | BoundsNotResolvedIncompleteGraph

type Function = {
    address: uint64
    mutable status: FunctionStatus
}

module JumpTable =
    type T = {
        reference: uint64
        address: uint64
        mutable firstIndex: int
        mutable count: int
    }

    let create (reference: uint64) (address: uint64): T = {
        reference = reference
        address = address
        firstIndex = 0
        count = 0
    }

let private logger = LogManager.GetCurrentClassLogger()

type T = {
    graph: InstructionGraph.T
    importResolver: IImportResolver.IImportResolver
    functions: SortedDictionary<uint64, Function>
    jumpTables: SortedDictionary<uint64, JumpTable.T>
}

let create (textBytes: ArraySegment<byte>) (pc: uint64) (importResolver: IImportResolver.IImportResolver): T = {
    graph = InstructionGraph.disassemble textBytes pc
    importResolver = importResolver
    functions = new SortedDictionary<uint64, Function>()
    jumpTables = new SortedDictionary<uint64, JumpTable.T>()
}

let analyze (t: T): unit =
    addExplicitCalls t
    resolveFunctionBounds t
    resolveExternalFunctionCalls t

let getGraph ({ graph = graph }): InstructionGraph.T = graph

let addFunction (t: T) (address: uint64): unit =
    t.functions.Add(address, { address = address; status = Created })

let private addExplicitCalls (t: T): unit =
    logger.Info("Adding explicit calls")
    for instr in t.graph.Instructions do
        if instr.Code = MnemonicCode.Icall && instr.Operands.[0].Type = OperandType.ImmediateBranch then
            let targetAddress = instr.GetTargetAddress()
            t.graph.AddLink(instr.Offset, targetAddress, InstructionGraph.LinkType.Call)
            if not (t.functions.ContainsKey(targetAddress)) then
                t.functions.Add(targetAddress, { address = targetAddress; status = Created })
    logger.Info("Done")

let private resolveFunctionBounds (t: T): unit =
    logger.Info("Resolving function bounds")
    let mutable index = 0
    let keys = t.functions.Keys
    let rec run (keys: uint64 list): unit =
        match keys with
        | key :: rest ->
            let func = t.functions.[key]
            index <- index + 1
            logger.Debug("[{0}/{1}] sub_{2:x8}...", index, t.functions.Count, func.address)
            if not (t.graph.InBounds(func.address)) then
                logger.Warn("The specified address is outside of code segment");
                func.status <- BoundsNotResolvedInvalidAddress
                run rest
            else
                if not (t.graph.ContainsAddress(func.address)) then
                    logger.Debug("The specified address is not a valid start of instruction, re-disassembling");
                    t.graph.Redisassemble(func.address);
                let visitedAllLinks =
                    t.graph.AsGenericGraph()
                        |> Graph.withEdgeFilter (fun e -> (e.type_ &&& InstructionGraph.LinkType.Next ||| InstructionGraph.LinkType.Branch ||| InstructionGraph.LinkType.SwitchCaseJump) <> InstructionGraph.LinkType.None)
                        |> Graph.dfsWith func.address (fun instr _ ->
                            t.graph.GetExtraData(instr.Offset).functionAddress <- func.address
                            if instr.Code = MnemonicCode.Iret then
                                false
                            else
                                addNextLinks t instr
                                addExplicitBranches t instr
                                addSwitchCases t instr
                                true
                        )
                func.status <- if visitedAllLinks then BoundsResolved else BoundsNotResolvedIncompleteGraph
        | [] -> ()
    run (keys |> Seq.toList)
    logger.Info("Done")

let private resolveExternalFunctionCalls (t: T): unit =
    logger.Info("Resolving external calls")
    for instr in t.graph.Instructions do
        let import =
            (tryGetImportAddress t instr 1)
            |> Option.orElse (tryGetImportAddress t instr 0)
            |> Option.map (fun i -> t.importResolver.GetImportName i)
        match import with
        | Some name ->
            t.graph.GetExtraData(instr.Offset).importName <- name
        | None -> ()
    logger.Info("Done")

let private tryGetImportAddress (t: T) (instr: Instruction) (index: int): uint64 option =
    if instr.Operands.Count > index && instr.Operands.[index].Type = OperandType.Memory &&
        instr.Operands.[index].Base = OperandType.None && instr.Operands.[index].Index = OperandType.None &&
        t.importResolver.IsImportAddress((uint64)instr.Operands.[index].LValue.udword) then
        Some instr.Operands.[index].LValue.uqword
    else
        None

let private addNextLinks (t: T) (instr: Instruction): unit =
    match instr.Code with
        | MnemonicCode.Iret
        | MnemonicCode.Ijmp
        | MnemonicCode.Iint3 -> ()
        | _ ->
            let next = t.graph.GetNext(instr.Offset)
            if next <> t.graph.FirstAddressAfterCode then
                t.graph.AddLink(instr.Offset, next, InstructionGraph.LinkType.Next)

let private addExplicitBranches (t: T) (instr: Instruction): unit =
    match instr.Code with
        | MnemonicCode.Ija
        | MnemonicCode.Ijae
        | MnemonicCode.Ijb
        | MnemonicCode.Ijbe
        | MnemonicCode.Ijcxz
        | MnemonicCode.Ijecxz
        | MnemonicCode.Ijg
        | MnemonicCode.Ijge
        | MnemonicCode.Ijl
        | MnemonicCode.Ijle
        | MnemonicCode.Ijmp
        | MnemonicCode.Ijno
        | MnemonicCode.Ijnp
        | MnemonicCode.Ijns
        | MnemonicCode.Ijnz
        | MnemonicCode.Ijo
        | MnemonicCode.Ijp
        | MnemonicCode.Ijrcxz
        | MnemonicCode.Ijs
        | MnemonicCode.Ijz ->
            if instr.Operands.[0].Type = OperandType.ImmediateBranch then
                t.graph.AddLink(instr.Offset, instr.GetTargetAddress(), InstructionGraph.LinkType.Branch)
        | _ -> ()

let private addSwitchCases (t: T) (instr: Instruction): unit =
    if instr.Code = MnemonicCode.Ijmp && instr.Operands.[0].Type = OperandType.Memory then
        let address = (uint64)instr.Operands.[0].LValue.udword
        let table =
            match t.jumpTables.TryGetValue(address) with
            | (true, table) -> table
            | (false, _) ->
                let table = JumpTable.create instr.Offset address
                resolveJumpTable t table
                t.jumpTables.Add(address, table)
                table
        for i in [table.firstIndex .. table.count - 1] do
            t.graph.AddLink(table.reference, t.graph.ReadUInt32(table.address + (uint64)(i * 4)), InstructionGraph.LinkType.SwitchCaseJump)

let private resolveJumpTable (t: T) (table: JumpTable.T): unit =
    if table.address >= t.graph.FirstAddressAfterCode then
        logger.Warn("Jump table is not in the code segment")
    else
        logger.Info("Resolving jump table tbl_{0:x8}", table.address)
        let mutable idx = OperandType.None
        let mutable lowByteIdx = OperandType.None
        let mutable indirectAddress = 0uL
        let casesCountOption =
            t.graph.AsGenericGraph()
                |> Graph.withEdgeFilter (fun e -> (e.type_ &&& InstructionGraph.LinkType.Next ||| InstructionGraph.LinkType.Branch) <> InstructionGraph.LinkType.None)
                |> Graph.reverseEdges
                |> Graph.dfsPick table.reference (fun instr link ->
                    // find out the jump index register
                    if idx = OperandType.None then
                        idx <- instr.Operands.[0].Index
                        lowByteIdx <- getLowByteRegisterFromDword idx
                        Continue
                    else
                        // update the main jump register if the jump is indirect
                        if indirectAddress = 0uL && instr.Code = MnemonicCode.Imov && instr.Operands.[0].Base = lowByteIdx then
                            idx <- instr.Operands.[1].Base
                            indirectAddress <- (uint64)instr.Operands.[1].LValue.udword
                            Continue
                        else
                            // search for the jump to the default case
                            if not (instr.Code = MnemonicCode.Ija && instr.Operands.[0].Type = OperandType.ImmediateBranch) then
                                Continue
                            else
                                // search for cases count, can find it from code like cmp ecx, 0xb
                                // the jump register is irrelevant since it must be the closest one to ja
                                let value = AssignmentTracker.find t.graph instr.Offset idx (fun i _ ->
                                    i.Code = MnemonicCode.Icmp && i.Operands.[0].Type = OperandType.Register)
                                Return (value |> Option.map (fun v -> (int)v.ubyte + 1))
                )

        match casesCountOption with
        | Some casesCount ->
            let mutable (jumpsCount, casesCount) =
                if indirectAddress = 0uL then
                    (casesCount, 0)
                else
                    ((int)(t.graph.GetBytes(indirectAddress, casesCount).Max()) + 1, casesCount)

            let mutable offset = 0uL
            for i in [0 .. jumpsCount - 1] do
                if not (t.graph.AddJumpTableEntry(table.address + offset)) then
                    table.firstIndex <- table.firstIndex + 1
                t.graph.AddJumpTableEntry(table.address + offset) |> ignore
                offset <- offset + 4uL
            table.count <- jumpsCount
            while casesCount >= 4 do
                t.graph.AddJumpTableIndirectEntries(table.address + offset, 4)
                casesCount <- casesCount - 4
                offset <- offset + 4uL
            if casesCount > 0 then
                t.graph.AddJumpTableIndirectEntries(table.address + offset, casesCount)
                offset <- offset + (uint64)casesCount
            t.graph.Redisassemble(table.address + offset)
        | None -> ()
        logger.Info("Done")

let private getLowByteRegisterFromDword (register: OperandType): OperandType =
    match register with
        | OperandType.EAX -> OperandType.AL
        | OperandType.EBX -> OperandType.BL
        | OperandType.ECX -> OperandType.CL
        | OperandType.EDX -> OperandType.DL
        | _ -> OperandType.None
